pkg/binary/envoy/debug/pii_test.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2019 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package debug
+
+import (
+	"testing"
+
+	"bitbucket.org/creachadair/shell"
+	"github.com/google/go-cmp/cmp"
+)
+
+func TestProcess(t *testing.T) {
+	format := "[%START_TIME%] \"%REQ(:METHOD)% %REQ(X-ENVOY-ORIGINAL-PATH?:PATH)%" +
+		" %PROTOCOL%\" %RESPONSE_CODE% %RESPONSE_FLAGS% \"%DYNAMIC_METADATA(istio.mixer:status)%\"" +
+		" \"%UPSTREAM_TRANSPORT_FAILURE_REASON%\" %BYTES_RECEIVED% %BYTES_SENT% %DURATION% %RESP(X-ENVOY-UPSTREAM-SERVICE-TIME)%" +
+		" \"%REQ(X-FORWARDED-FOR)%\" \"%REQ(USER-AGENT)%\" \"%REQ(X-REQUEST-ID)%\" \"%REQ(:AUTHORITY)%\" \"%UPSTREAM_HOST%\"" +
+		" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_LOCAL_ADDRESS% %DOWNSTREAM_REMOTE_ADDRESS% %REQUESTED_SERVER_NAME%\n"
+
+	formats, ok := shell.Split(format)
+	if !ok {
+		t.Fatalf("error in splitting the format: %s", format)
+	}
+
+	validLog := "[2019-09-05T17:53:36.908Z] \"GET /productpage HTTP/1.1\" 200" +
+		" - \" - \" \" - \" 0 5183 47 46 \"10.28.2.1\" \"Mozilla/5.0 (X11; Linux x86_64)" +
+		" AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36\"" +
+		" \"4273c09c-2534-962a-b7da-04a04cf39e88\" \"35.199.144.198\" \"127.0.0.1:9080\"" +
+		" inbound|9080|http|productpage.default.svc.cluster.local - 10.28.2.27:9080 10.28.2.1:0" +
+		" outbound_.9080_._.productpage.default.svc.cluster.local"
+	logWithOneField := `[2019-09-05T17:53:36.908Z]`
+
+	type Out struct {
+		logs []string
+		err  error
+	}
+	// test cases wrapped up in a map where key is the test case name and value is the struct
+	// defining the inputs and expecte outcome
+	tests := map[string]struct {
+		logs        []string
+		format      string
+		containsPII map[string]bool
+		want        Out
```

Comment:
Don't do this, it adds a ton of overhead when you're building the test case. Instead just flatten it out (you can do something simple like put a newline in the struct definition, or a comment like `// output fields` separating the input data from output):
```go
tests := map[string]struct {
	logs   []string
	format string
	pii    map[string]bool

	outLogs []string
	err     string
}{
	"simple": {
		logs: []string{validLog, logWithOneField},
		format: format,
		pii: map[string]bool{"[%START_TIME%]": true},
		outLogs: []string{validLog},
		err: "",
	},
}
```
This is much clearer to read. Some other things to note here:
- Try to keep your field initializations consistent. Don't mix with one field on one line and two on another, etc. Either initialize the entire thing on one line (`logs: []string{validLog, logWithOneField},`) or have exactly one field per line (the entire `"simple"` object above).
- The closing bracket always goes on a new line
- Always leave a trailing comma when you're initializing a struct (that last `err: nil,`). You don't have to with a list literal (`[]string{validLog, logWithOneField}` is correct, avoid `[]string{validLog, logWithOneField,}`)
- We can simplify the name `containsPII` - we know in context what it is / how it's used, so we don't need a long name. Just calling it `pii` saves us quite a bit of visual space and doesn't lose us any understanding.
- You almost never want an `error` in your test case struct, as building a correct error is a real pain. Instead, you should use a string and a `string.Contains(case.error, actualError.Error())` to compare. Then you can just ensure that the text for the expected exit condition is being hit without having to build an entire `error` object. Everywhere that you have `case.err != nil` just becomes `case.err != ""`